module/VuFind/src/VuFind/Db/Service/OaiResumptionService.php

```diff
@@ -0,0 +1,153 @@
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Sudharma Kellampalli <skellamp@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+
+namespace VuFind\Db\Service;
+
+use Laminas\Log\LoggerAwareInterface;
+use VuFind\Db\Entity\OaiResumption;
+use VuFind\Log\LoggerAwareTrait;
+
+/**
+ * Database service for OaiResumption.
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Sudharma Kellampalli <skellamp@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+class OaiResumptionService extends AbstractService implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    /**
+     * Remove all expired tokens from the database.
+     *
+     * @return void
+     */
+    public function removeExpired()
+    {
+        $dql = "DELETE FROM " . $this->getEntityClass(OaiResumption::class) . " O "
+            . "WHERE O.expires <= :now";
+        $parameters['now'] = new \DateTime();
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $query->execute();
+    }
+
+    /**
+     * Retrieve a row from the database based on primary key; return null if it
+     * is not found.
+     *
+     * @param string $token The resumption token to retrieve.
+     *
+     * @return OaiResumption|null
+     */
+    public function findToken($token)
+    {
+        $dql = "SELECT O "
+        . "FROM " . $this->getEntityClass(OaiResumption::class) . " O "
+        . "WHERE O.id = :token";
+        $parameters = compact('token');
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $records = $query->getResult();
+        return current($records);
+    }
+
+    /**
+     * Create a new resumption token
+     *
+     * @param array $params Parameters associated with the token.
+     * @param int   $expire Expiration time for token (Unix timestamp).
+     *
+     * @return int          ID of new token
+     */
+    public function saveToken($params, $expire)
+    {
+        $row = $this->createEntity()
+            ->setResumptionParameters($this->saveParams($params))
+            ->setExpiry(\DateTime::createFromFormat('U', $expire));
+        try {
+            $this->persistEntity($row);
+        } catch (\Exception $e) {
+            $this->logError('Could not save token: ' . $e->getMessage());
+            return false;
+        }
+        return $row->getId();
+    }
+
+    /**
+     * Extract an array of parameters from the object.
+     *
+     * @param OaiResumption $oaiEntity OaiResumption object
+     *
+     * @return array Original saved parameters.
+     */
+    public function restoreParams($oaiEntity)
```

Comment:
If my suggestion of replacing logic below with `http_build_query` is successful, then I think we can also completely eliminate this method and instead make a call to `parse_str` in the code that currently uses this.